Source/DataStructureAndEncodingDefinition/fixed_istream_buffer.h

```diff
@@ -0,0 +1,124 @@
+        : stream(stream), size(size), half(size / 2), buffer(size), base() {
+        std::streamsize read = stream.read(buffer.data(), size).gcount();
+        setg(buffer.data(), buffer.data(), buffer.data() + read);
+    }
+
+    int underflow() override {
+        if (gptr() == buffer.data() + size) {
+            char* half_buffer = eback() + half;
+            std::memmove(eback(), half_buffer, half);
+            std::streamsize read = stream.read(half_buffer, half).gcount();
+            setg(eback(), half_buffer, half_buffer + read);
+            base += half;
+        }
+        return has_reached_end() ? traits_type::eof() : traits_type::to_int_type(*gptr());
+    }
+
+    std::streampos seekoff(off_type offset, std::ios_base::seekdir way, std::ios_base::openmode which) override {
+        if(input_sequence(which)) {
+            bool after_beginning = buffer_size_read() >= -offset;
+            bool before_end = buffer_size_unread() > offset;
+            if(way == std::ios_base::cur) {
+                if(after_beginning && before_end) {
+                  gbump(offset);
+                  return pos_type(base + buffer_size_read());
+                } else if (after_beginning) {
+                  return seek_offset_forward_via_read(offset);
+                }
+            } else if (way == std::ios_base::beg && offset >= 0) {
+                return seekpos(pos_type(offset), which);
+            }
+        }
+        return invalid_position();
+    }
+
+    std::streampos seekpos(pos_type pos, std::ios_base::openmode which) override {
+        bool after_beginning = base <= off_type(pos);
+        bool before_end = base + buffer_size_seekable() > off_type(pos);
+        if(input_sequence(which)) {
+            if(after_beginning && before_end) {
+                setg(eback(), eback() + (off_type(pos) - base), egptr());
+                return pos_type(base + buffer_size_read());
+            } else if(after_beginning) {
+                return seek_position_forward_via_read(pos);
+            }
+        }
+        return invalid_position();
+    }
+
+    std::streampos seek_offset_forward_via_read(off_type offset) {
+        std::streampos seekpos = base + buffer_size_read() + offset;
+        return seek_position_forward_via_read(seekpos);
+    }
+
+    // this is potentially expensive, especially when reading immediatley after seeking
+    // in the usual case this will not be called
+    // assumes seekpos is past the end of the current buffer
+    std::streampos seek_position_forward_via_read(std::streampos seekpos) {
+        std::streampos endpos = base + buffer_size_seekable() - 1;
+        unsigned total_read_length = seekpos - endpos;
+        unsigned full_reads = total_read_length / size;
+        for(unsigned i = 0; i < full_reads; ++i) {
+          std::streamsize read = stream.read(buffer.data(), size).gcount();
+          setg(buffer.data(), buffer.data() + read - 1, buffer.data() + read);
```

Comment:
does this need to be guarded against a read of 0? not sure if the g pointer needs to be valid on a failed operation.